File: leetcode/1254. Number of Closed Islands.py
class Solution:
    def DFS(self, x, y, grid):
        # 邊界檢查
        # 確定是在範圍內
        # if x < 0 or x > self.x_len - 1 or y < 0 or y > self.y_len - 1:
        #     return
        self.x_move = [1, -1, 0, 0]
        self.y_move = [0, 0, 1, -1]
        
        # 將目前檢查的格子標記為已經檢查過
        grid[x][y] = 1
        # 檢查四個方向
        for i in range(4):
            tx = x + self.x_move[i]
            ty = y + self.y_move[i]
            # 如果超出範圍或是遇到海洋，則跳過
            if tx < 0 or tx > self.x_len - 1 or ty < 0 or ty > self.y_len - 1 or grid[tx][ty] == 1:
                continue
            self.DFS(tx, ty, grid)
    
    def closedIsland(self, grid: List[List[int]]) -> int:
        # 計算長寬
        self.x_len = len(grid)
        self.y_len = len(grid[0])
        # 移除邊界的陸地及其相連的陸地
        # 這樣就只剩下孤島了
        for x in range(self.x_len):
            for y in range(self.y_len):
                if (x == 0 or y == 0 or x == self.x_len - 1 or y == self.y_len - 1) and grid[x][y] == 0:
                    # 在做DFS時會將相連的陸地標記為海洋(也就是檢查過)
                    self.DFS(x, y, grid)
        count = 0
        # 計算孤島的數量
        for x in range(self.x_len):
            for y in range(self.y_len):
                if grid[x][y] == 0:
                    count += 1
                    self.DFS(x, y, grid)
        return count

File: leetcode/1254. Number of Closed Islands.cpp
class Solution {
public:

    int x_len, y_len;
    int x_move[4] = {1, -1, 0, 0},
        y_move[4] = {0, 0, 1, -1};

    void DFS(int x, int y, vector<vector<int> >& grid){
        // 保證grid[x][y]為陸地
        grid[x][y] = 1;
        for (int i=0; i<4; i++){
            int tx = x + x_move[i],
                ty = y + y_move[i];
            if (tx >= 0 and tx < x_len and ty >= 0 and ty < y_len and grid[tx][ty] == 0)
                DFS(tx, ty, grid);
        }
    }

    int closedIsland(vector<vector<int>>& grid) {
        // 計算二維陣列長寬
        x_len = grid.size();
        y_len = grid[0].size();
        // 移除邊界的陸地及其相連的陸地
        for (int x=0; x<x_len; x++){
            for (int y=0; y<y_len; y++){
                if ((x == 0 or y == 0 or x == x_len-1 or y == y_len-1) and grid[x][y] == 0)
                    DFS(x, y, grid);
            }
        }
        // 計算剩餘的島嶼數量
        int ans = 0;
        for (int x=0; x<x_len; x++){
            for (int y=0; y<y_len; y++){
                if (grid[x][y] == 0){
                    ans++;
                    DFS(x, y, grid);
                }
            }
        }
        return ans;

    }
};